Fix groups after the player model refactor

maximmaxim345 opened the pull request:
Fixes around player group support after the large player model refactor.
For a list of all changes and their reasons see the commits.

Frontend fixes are in:
- https://github.com/music-assistant/frontend/pull/1123

Issues left:
- [x] Leaders in `SyncGroupPlayer`s still have the groups queue after powering it off
- [x] `SyncGroupPlayer.supported_features` is not set based on the group members capabilities.
- [x] Switch leader when it gets ungrouped
- [x] Support empty sync groups
- [x] Multiple groups can be activated simultaneously
- [x] `update_state` that doesn't update `synced_to` of other players in some cases
- [x] Groups don't always get powered off it conflicts with another group
- [x] There still a couple of delays waiting for state updates, but they are pretty short now.
- [x] Backup/restore logic on restart of Music Assistant 

Future work (for future PRs):
- Properly support all `supported_features` of the underlying player within `SyncGroupPlayer`.
- State updates sometimes happen multiple times per player (in `signal_player_state_update`).
- Rename `active_group` to `grouped_to` or consolidate `synced_to` and `active_group` to just `grouped_to`
- add a `static_members` to show static group members in the frontend

maximmaxim345 committed:
fix: copy group members value to avoid mutations

This fixes an issue where add player automatically become static group
members since the reference to the list of group members and static
group members were the same.

maximmaxim345 committed:
fix(ugp): add missing polling for universal groups

Polling was also enabled before the refactor and is required to update
the state.

maximmaxim345 committed:
fix(ugp): call `update_state` once we are done updating

maximmaxim345 committed:
fix(ugp): set initial group members only after the player was registered

Since `self.config` is not fully initialized until the player was
registered. Before `group_members` was always `[]` on start.

maximmaxim345 committed:
fix(ugp): remove double call to `update_state()`

`_set_attributes()` already calls that.

maximmaxim345 committed:
fix(ugp): delete the stream after stopping

maximmaxim345 committed:
fix(ugp): immediately update the state after stopping the universal group

maximmaxim345 committed:
fix(ugp): set active_source when starting playback

maximmaxim345 committed:
fix(ugp): make it always available

Just as before the refactor.

maximmaxim345 committed:
docs(demo_music_provider): wrong example text

maximmaxim345 committed:
fix(demo_player_provider): Add group player features

maximmaxim345 committed:
refactor: add `Player.on_registered`

For handling additional initialization that requires `self.config`.

maximmaxim345 committed:
fix(SyncGroupPlayer): load group members list after registration

Since `self.config` is not fully loaded in `__init__`.

maximmaxim345 committed:
docs(demo_player_provider): add docs for `on_registered`

marcelveldt commented on music_assistant/models/player.py:
```diff
@@ -633,13 +643,17 @@ def synced_to(self) -> str | None:
 
         If this player is not synced to another player (or is the sync leader itself),
         this should return None.
+        If this player is a member of a active permanent group, this should return the id of the
+        group.
```
No, that is not true. This can only return the ID of a"sync leader" - a group is not that

maximmaxim345 replied:
Ok, I'll adjust the default implementation to ignore group players then.

marcelveldt replied:
You should just override it in the group player.

marcelveldt replied:
synced_to (as the name implies!) can only return an id of a player to which a player is synced to, so not an id of a player the player is part of the group. We have "active_group" for that

marcelveldt commented on music_assistant/models/player.py:
```diff
@@ -633,13 +643,17 @@ def synced_to(self) -> str | None:
-            if self.player_id in player.group_members:
+            if self.player_id in player.group_members and (
+                player.powered or player.active_group == PlaybackState.PLAYING
+            ):
```
why this extra check ?

maximmaxim345 replied:
Since the refactor the default implementation of `synced_to` always pointed to a group player if there was one.
So the frontend hid all players that were part of the group, no matter if it was powered on or not.
But that won't be needed when I fix `synced_to`.

Wait..., `player.active_group == PlaybackState.PLAYING` makes no sense...

marcelveldt replied:
yeah, you can leave this out

marcelveldt commented on music_assistant/models/player.py:
```diff
@@ -1332,10 +1346,17 @@ def __init__(
 
+    async def on_registered(self) -> None:
+        """Complete the initialization once the player was registered."""
+        # Config is only available after the player was registered
+        self._attr_group_members = list(
+            cast("list[str]", self.config.get_value(CONF_GROUP_MEMBERS, []))
+        )
```
I'm not fully understanding why this is different then setting it in the init of the player ?

maximmaxim345 replied:
`self.config` is empty at `__init__`. The full config is only loaded after it was registered [here](https://github.com/music-assistant/server/blob/087e36d55e1c8d9c054547ae1e48dc936821cf2b/music_assistant/controllers/players.py#L1186)
Before that `self.config` is pretty much empty, so `group_members` is always empty. See [here](https://github.com/music-assistant/server/blob/087e36d55e1c8d9c054547ae1e48dc936821cf2b/music_assistant/models/player.py#L153C9-L153C92) and [here](https://github.com/music-assistant/server/blob/ebb843a45329fe3455161702987c8e8798ca60e1/music_assistant/controllers/config.py#L389C1-L401C48).

marcelveldt commented on music_assistant/providers/universal_group/player.py:
```diff
@@ -162,7 +171,7 @@ async def power(self, powered: bool) -> None:
                         await asyncio.sleep(1)
                     await asyncio.sleep(1)
-                if member.synced_to:
+                if member.synced_to and member.synced_to != self.player_id:
```
This can not ever be true so this check is wrong

marcelveldt commented on music_assistant/providers/universal_group/player.py:
```diff
@@ -241,6 +251,7 @@ async def play_media(self, media: PlayerMedia) -> None:
         self._attr_elapsed_time = 0
         self._attr_elapsed_time_last_updated = time() - 1
         self._attr_playback_state = PlaybackState.PLAYING
+        self._attr_active_source = media.queue_id
```
Reading this I think we should have named "queue_id" in the PlayerMedia "source_id" as this is now super confusing. Let's adjust that in the models and add a fallback property for backwards compatibility

marcelveldt replied:
I have made this adjustment in the models - will merge after this PR

maximmaxim345 committed:
fix: set `synced_to` to `None` for group players

maximmaxim345 committed:
fix: set `synced_to` to `None` for group members

maximmaxim345 committed:
fix(SyncGroupPlayer): forward playback state and elapsed time for `SyncGroupPlayers`

maximmaxim345 committed:
fix(SyncGroupPlayer): set `current_media` and `active_source` in `play_media`

maximmaxim345 committed:
fix: also update active group on member updates

maximmaxim345 committed:
docs: clarify `synced_to` documentation for groups

maximmaxim345 committed:
docs(SyncGroupPlayer): shorten doc comment

maximmaxim345 committed:
fix(ugp): don't allow universal groups to group other groups

maximmaxim345 committed:
fix: disallow grouping of groups

maximmaxim345 committed:
fix(SyncGroupPlayer): power off handling was never handled

maximmaxim345 committed:
fix(SyncGroupPlayer): restore leaders queue after powering the group off

maximmaxim345 committed:
fix(SyncGroupPlayer): only sync playback state when the group is powered on

maximmaxim345 committed:
refactor: add setters for `active_source` and `current_media`

Since both properties can be overwritten, setting `_attr_active_source`
and `_attr_current_media` is not guaranteed to work.

maximmaxim345 committed:
fix(SyncGroupPlayer): call `_set_attributes` only once `self.config` is available

maximmaxim345 committed:
fix(SyncGroupPlayer): set `can_group_with` to compatible players

maximmaxim345 committed:
fix(SyncGroupPlayer): also forward `supported_features` and `flow_mode`

maximmaxim345 committed:
fix(api): add edge cases when grouping already grouped/synced players

maximmaxim345 committed:
fix(SyncGroupPlayer): allow empty groups and leader changes

maximmaxim345 committed:
fix(SyncGroupPlayer): add edge case when all group members get removed

maximmaxim345 committed:
fix(SyncGroupPlayer): transfer playback on queue changes

maximmaxim345 committed:
fix(SyncGroupPlayer): raise when playing to an empty group

maximmaxim345 committed:
refactor(SyncGroupPlayer): rename `save` to `backup`

maximmaxim345 committed:
fix(SyncGroupPlayer): try to power off the other group if ungrouping fails

maximmaxim345 committed:
fix(SyncGroupPlayer): ungroup synced players before adding to the player group

maximmaxim345 committed:
fix(SyncGroupPlayer): only setup sync group once we handled the groups members

maximmaxim345 committed:
refactor(SyncGroupPlayer): add `_handle_member_collisions`

maximmaxim345 committed:
refactor(SyncGroupPlayer): also handle group collisions in `set_members`

maximmaxim345 committed:
fix: typos

maximmaxim345 committed:
fix(SyncGroupPlayer): init `sync_leader` earlier

maximmaxim345 committed:
refactor(SyncGroupPlayer): call player commands directly

maximmaxim345 committed:
refactor(SyncGroupPlayer): add helpers for dissolving leader transition

maximmaxim345 committed:
fix(SyncGroupPlayer): `can_group_with` for empty groups

maximmaxim345 committed:
fix(SyncGroupPlayer): leader transition

maximmaxim345 committed:
fix(SyncGroupPlayer): wait for state update of removed members

maximmaxim345 committed:
feat: add `Player.active_groups` to get all active groups

And not only the first one.

maximmaxim345 committed:
fix(SyncGroupPlayer): fix collision logic for multiple active groups

maximmaxim345 committed:
refactor(SyncGroupPlayer): inline `_set_attributes()`

maximmaxim345 committed:
fix(SyncGroupPlayer): disable all other features than the basic ones for now

maximmaxim345 committed:
fix(SyncGroupPlayer): reduce sleep delay to 0.1s

maximmaxim345 commented on music_assistant/models/player.py:
```diff
@@ -1409,60 +1512,126 @@ async def pause(self) -> None:
+                    # member, we need to power it off to leave the group
+                    await other_group.power(False)
+                # Wait a bit so the state can settle
+                await asyncio.sleep(0.1)
```
is it fine to have short sleeps here?
The reason why this is required here is that:
- After the refactor properties like `synced_to` and `active_group`  are computed based on other players
- calling `update_state` doesn't call `update_state` of other players directly, but only through `self.mass.loop.call_soon`
- so if the player is updated, those properties are not yet correct for all players

Or should I change the logic to be able to handle incorrect states?
Or is it better to make `update_state` update all states immediately instead?

marcelveldt replied:
I think its better to simply call update_state directly on the child players to prevent race conditions like this. In the end a sleep is also just a bit of guessing

maximmaxim345 replied:
Done

maximmaxim345 commented on music_assistant/models/player.py:
```diff
@@ -1409,60 +1512,126 @@ async def pause(self) -> None:
+            self._leader_backup_current_media = None
+
+    def _restore_leader_queue(self) -> None:
+        if leader := self.sync_leader:
```
I admit, this is pretty hacky. Do you have an idea how to better solve/refactor this?
I think this workaround is also not working across restarts of Music Assistant.

maximmaxim345 commented on music_assistant/models/player.py:
```diff
@@ -1409,60 +1512,126 @@ async def pause(self) -> None:
             )
+            self.sync_leader = None
+
+    def _backup_leader_queue(self) -> None:
+        if leader := self.sync_leader:
+            self._leader_backup_active_source = leader.active_source
+            self._leader_backup_current_media = leader.current_media
+        else:
+            self._leader_backup_active_source = None
+            self._leader_backup_current_media = None
+
+    def _restore_leader_queue(self) -> None:
+        if leader := self.sync_leader:
+            leader.active_source = self._leader_backup_active_source
+            leader.current_media = self._leader_backup_current_media
+        self._leader_backup_active_source = None
+        self._leader_backup_current_media = None
```
I admit, this is pretty hacky. Do you have an idea how to better solve/refactor this?
I think this workaround is also not working across restarts of Music Assistant.

marcelveldt replied:
why would you need this ?

marcelveldt replied:
imo you can simply restore active source and current media to None and the provider implementation will override it if it knows better.

maximmaxim345 replied:
~If you turn off a group (or ungroup the leader), the leader still thinks it is playing the groups stream. The frontend for example then thinks the powered off group is playing for some reason.
This restores whatever was playing before that, so turning on and off a group won't null out what was playing before.~

Edit: Will try that out

maximmaxim345 replied:
Works!

maximmaxim345 commented:
After the refactor static member are now members that are always member to that group, as opposed to members that are by default part of that group (like it was before).
One advantage right now is that regular group members and static group members are not differentiated in the frontend UI. 
Meaning: static members are still shown with an checkbox, but clicking it just throws an error message.

maximmaxim345 committed:
fix(SyncGroupPlayer): cache `active_groups`

maximmaxim345 commented on music_assistant/models/player.py:
```diff
@@ -1409,60 +1512,126 @@ async def pause(self) -> None:
+                # Wait a bit so the state can settle
+                await asyncio.sleep(0.1)
+        if (
+            member.synced_to is not None
```
The code is only called when `sync_leader` is set (and not None): If we have group members that we need to handle, we also have a sync leader.

maximmaxim345 committed:
fix: update children immediately with `update_state`

maximmaxim345 committed:
fix(SyncGroupPlayer): reset leader queue instead of hacky backup/restore logic

maximmaxim345 committed:
Merge branch 'dev' into fix/groups-after-refactor

marcelveldt commented on music_assistant/controllers/players.py:
```diff
@@ -1385,13 +1423,18 @@ def signal_player_state_update(
         # update/signal group player(s) child's when group updates
         for child_player in self.iter_group_members(player, exclude_self=True):
-            self.mass.loop.call_soon(child_player.update_state, True)
+            child_player.update_state(True)
```
why force_update set to True ? That will create a redundant event

marcelveldt replied:
```suggestion
            child_player.update_state()
```

marcelveldt commented on music_assistant/controllers/players.py:
```diff
@@ -1385,13 +1423,18 @@ def signal_player_state_update(
         # update/signal group player(s) when child updates
         for group_player in self._get_player_groups(player, powered_only=False):
-            self.mass.loop.call_soon(group_player.update_state, True)
+            group_player.update_state(True)
```
```suggestion
            group_player.update_state()
```

marcelveldt commented on music_assistant/controllers/players.py:
```diff
@@ -1385,13 +1423,18 @@ def signal_player_state_update(
         # update/signal manually synced to player when child updates
         if (synced_to := player.synced_to) and (synced_to_player := self.get(synced_to)):
-            self.mass.loop.call_soon(synced_to_player.update_state, True)
+            synced_to_player.update_state(True)
```
```suggestion
            synced_to_player.update_state()
```

marcelveldt commented on music_assistant/controllers/players.py:
```diff
@@ -1385,13 +1423,18 @@ def signal_player_state_update(
+        if (active_group := player.active_group) and (
+            active_group_player := self.get(active_group)
+        ):
+            active_group_player.update_state(True)
```
```suggestion
            active_group_player.update_state()
```

maximmaxim345 committed:
Apply suggestion from @marcelveldt

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

maximmaxim345 committed:
Apply suggestion from @marcelveldt

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

maximmaxim345 committed:
Apply suggestion from @marcelveldt

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

maximmaxim345 committed:
Apply suggestion from @marcelveldt

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

marcelveldt commented on music_assistant/controllers/players.py:
```diff
@@ -1375,7 +1413,7 @@ def signal_player_state_update(
             for player_id in removed_members:
                 if removed_player := self.get(player_id):
-                    self.mass.loop.call_soon(removed_player.update_state, True)
+                    removed_player.update_state(True)
```
```suggestion
                    removed_player.update_state()
```

marcelveldt committed:
Update music_assistant/controllers/players.py

marcelveldt reviewed (approved):
Looks good! Thanks @maximmaxim345 !
